File: src/main/java/org/taktik/icure/logic/PrincipalLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * Foobar is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import org.taktik.icure.entities.Property;
import org.taktik.icure.entities.Role;
import org.taktik.icure.entities.base.Principal;
import org.taktik.icure.entities.embed.Permission;

import java.util.Set;

public interface PrincipalLogic<P extends Principal> {
	P getPrincipal(String principalId);

	Set<Property> getProperties(String principalId, boolean includeDirect, boolean includeHerited, boolean includeDefault);

	Set<Permission> getPermissions(String principalId, String virtualHostId, boolean includeDirect, boolean includeHerited, boolean includeDefault);

	Set<Role> getAscendantRoles(String principalId);
}

File: src/main/kotlin/org/taktik/icure/asynclogic/PrincipalLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.icure.entities.Role
import org.taktik.icure.entities.base.PropertyStub
import org.taktik.icure.entities.security.Permission
import org.taktik.icure.entities.security.Principal

interface PrincipalLogic<P : Principal> {
	suspend fun getPrincipal(principalId: String): P?
	fun getProperties(principalId: String, includeDirect: Boolean, includeHerited: Boolean, includeDefault: Boolean): Flow<PropertyStub>
	fun getPermissions(principalId: String, virtualHostId: String, includeDirect: Boolean, includeHerited: Boolean, includeDefault: Boolean): Flow<Permission>
	fun getAscendantRoles(principalId: String): Flow<Role>
}
